### PR TITLE
BGV: support user specifying the scheme param

### LINKED ISSUE #1416: Allow user to specify scheme parameter in input IR
```mlir
module attributes {scheme.bgv, bgv.schemeParam = #bgv.schemeParam<...>} {
  func.func @add(%arg0 : i16 {secret.secret}) {
    %1 = arith.add %arg0, %arg0 : i16
    return %1
  }
}
```

This can be user input as the frontend can easily generate such MLIR.

`--validate-noise` would then validate against such scheme parameter.

### PR DESCRIPTION
**Pull request by ZenithalHourlyRate**
Fixes #1416 

Waiting for #1400 and should be rebased after that being merged.

### TODO

Add tests

### CODE REVIEW

**Comment by ZenithalHourlyRate** [comment]
Rebased and ready for review as #1400 is merged.

**Review by j2kun — APPROVED** [review]

**Comment by j2kun** [comment]
This one is hitting an internal error so I will have to investigate a bit tomorrow

```
signal: illegal instruction
PLEASE submit a bug report to [http://go/llvm-crash-bug](https://www.google.com/url?q=http://go/llvm-crash-bug&sa=D) and include the crash backtrace.
Stack dump:
0.	Program arguments: heir-opt --mlir-to-secret-arithmetic --secret-insert-mgmt-bgv --validate-noise=model=bgv-noise-by-bound-coeff-average-case-pk /build/work/019369d93229b4f160c33e389e32f36b5527/google3/runfiles/google3/third_party/heir/tests/Transforms/validate_noise/validate_noise_preserve_user_param_fail.mlir --verify-diagnostics
#0 0x000055c8ca54bfab llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/build/cas/9d3/9d3ef63b65ac1e75fba3efda805d2d01583b6090b68f12f41a1eef9d85a0dfd3_020008d1b468+0x2d4bfab)
#1 0x000055c8ca549c1a llvm::sys::RunSignalHandlers() (/build/cas/9d3/9d3ef63b65ac1e75fba3efda805d2d01583b6090b68f12f41a1eef9d85a0dfd3_020008d1b468+0x2d49c1a)
#2 0x000055c8ca54c56b SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
#3 0x00007fcd34280e80 __restore_rt (/usr/grte/v5/lib64/libpthread.so.0+0x14e80)
#4 0x000055c8c92d1451 mlir::heir::bgv::NoiseByBoundCoeffModel<false, true>::evalModReduce(mlir::heir::bgv::LocalParam const&, mlir::heir::bgv::NoiseState const&) (/build/cas/9d3/9d3ef63b65ac1e75fba3efda805d2d01583b6090b68f12f41a1eef9d85a0dfd3_020008d1b468+0x1ad1451)
```

File: tests/Transforms/validate_noise/validate_noise_preserve_user_param_fail.mlir [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// RUN: heir-opt --mlir-to-secret-arithmetic --secret-insert-mgmt-bgv --validate-noise=model=bgv-noise-by-bound-coeff-average-case-pk %s --verify-diagnostics
+
+// expected-error@below {{'builtin.module' op Noise validation failed.}}
+module attributes {bgv.schemeParam = #bgv.scheme_param<logN = 11, Q = [17], P = [1093633], plaintextModulus = 65537>} {
```
Comment (j2kun) [inline tests/Transforms/validate_noise/validate_noise_preserve_user_param_fail.mlir:4]:
```suggestion
module attributes {bgv.schemeParam = #bgv.scheme_param<logN = 11, Q = [17, 18], P = [1093633], plaintextModulus = 65537>} {
```

I think the problem is here: the default scheme parameters generated in this case has only one logqi, but the [`getCurrentLevel` indexed here](https://github.com/google/heir/blob/fa56d4116a188911970b4a7fce8c7ebe7669d9b1/lib/Analysis/NoiseAnalysis/BGV/NoiseByBoundCoeffModel.cpp#L187) is 1. Adding a second entry to Q seems to fix the problem, but maybe you want to patch the code to be more robust to this edge case.

**Review by j2kun — CHANGES_REQUESTED** [review]

File: tests/Transforms/validate_noise/validate_noise_preserve_user_param_fail.mlir (hunk 1 above)
Comment (ZenithalHourlyRate) [reply tests/Transforms/validate_noise/validate_noise_preserve_user_param_fail.mlir:4]: OK I should then verify the user provided param has the same `Level` as the max level of the `MgmtAttr`. Should I push another version or open another PR?

Comment (ZenithalHourlyRate) [reply tests/Transforms/validate_noise/validate_noise_preserve_user_param_fail.mlir:4]: Also maybe we can add ASAN to the dev config/CI?

Comment (j2kun) [reply tests/Transforms/validate_noise/validate_noise_preserve_user_param_fail.mlir:4]: > OK I should then verify the user provided param has the same `Level` as the max level of the `MgmtAttr`. Should I push another version or open another PR?

You can amend the commit in this PR with your fix, it will propagate internally.

Comment (ZenithalHourlyRate) [reply tests/Transforms/validate_noise/validate_noise_preserve_user_param_fail.mlir:4]: pushed the fixed version with both fail noise validation test and fail level requirement test. In the future we should add more test/sanitizer on user input parameter.
